app/backend/src/couchers/jobs/handlers.py

```diff
@@ -40,3 +44,74 @@ def process_purge_signup_tokens(payload):
+
+
+def process_send_message_notifications(payload):
+    """
+    Sends out email notifications for messages that have been unseen for a long enough time
+    """
+    # very crude and dumb algorithm
+    logger.info(f"Sending out email notifications for unseen messages")
+
+    with session_scope() as session:
+        # users who have unnotified messages older than 30 minutes in any group chat
+        users = (
+            session.query(User)
+            .join(GroupChatSubscription, GroupChatSubscription.user_id == User.id)
+            .join(Message, Message.conversation_id == GroupChatSubscription.group_chat_id)
+            .filter(Message.time >= GroupChatSubscription.joined)
+            .filter(or_(Message.time <= GroupChatSubscription.left, GroupChatSubscription.left == None))
+            .filter(Message.id > User.last_notified_message_id)
+            .filter(Message.id > GroupChatSubscription.last_seen_message_id)
+            .filter(Message.time < now() - timedelta(minutes=30))
+            .filter(Message.message_type == MessageType.text)  # TODO: only text messages for now
+            .group_by(User)
+            .all()
+        )
+
+        for user in users:
+            # now actually grab all the group chats, not just less than 30 min old
+            subquery = (
+                session.query(
+                    GroupChatSubscription.group_chat_id.label("group_chat_id"),
+                    func.max(GroupChatSubscription.id).label("group_chat_subscriptions_id"),
+                    func.max(Message.id).label("message_id"),
+                    func.count(Message.id).label("count_unseen"),
+                )
+                .join(Message, Message.conversation_id == GroupChatSubscription.group_chat_id)
+                .filter(GroupChatSubscription.user_id == user.id)
+                .filter(Message.id > user.last_notified_message_id)
+                .filter(Message.id > GroupChatSubscription.last_seen_message_id)
+                .filter(Message.time >= GroupChatSubscription.joined)
+                .filter(Message.message_type == MessageType.text)  # TODO: only text messages for now
+                .filter(or_(Message.time <= GroupChatSubscription.left, GroupChatSubscription.left == None))
+                .group_by(GroupChatSubscription.group_chat_id)
+                .order_by(func.max(Message.id).desc())
+                .subquery()
+            )
+
+            unseen_messages = (
+                session.query(GroupChat, Message, subquery.c.count_unseen)
+                .join(subquery, subquery.c.message_id == Message.id)
+                .join(GroupChat, GroupChat.conversation_id == subquery.c.group_chat_id)
+                .order_by(subquery.c.message_id.desc())
+                .all()
+            )
+
+            user.last_notified_message_id = max([message.id for _, message, _ in unseen_messages])
```

Comment:
Drop the [ ], to instead pass a *generator expression* to `max`. Two characters less to write and doesn't create an intermediary array.